javalib/src/main/scala/java/util/HashMap.scala

```diff
@@ -12,153 +12,502 @@
+            // go to low
+            node.previous = lastLow
+            if (lastLow eq null)
+              newTable(i) = node
+            else
+              lastLow.next = node
+            lastLow = node
+          } else {
+            // go to high
+            node.previous = lastHigh
+            if (lastHigh eq null)
+              newTable(oldlen + i) = node
+            else
+              lastHigh.next = node
+            lastHigh = node
+          }
+          node = node.next
+        }
+        if (lastLow ne null)
+          lastLow.next = null
+        if (lastHigh ne null)
+          lastHigh.next = null
+        i += 1
       }
     }
   }
 
-  private class ValuesView extends AbstractMapView[V] {
-    override def size(): Int =
-      inner.size
+  private[this] def tableSizeFor(capacity: Int): Int =
+    Math.min(Integer.highestOneBit(Math.max(capacity - 1, 4)) * 2, 1 << 30)
+
+  private[this] def newThreshold(size: Int) =
+    (size.toDouble * loadFactor).toInt
+
+  // Iterators
+
+  private[util] def nodeIterator(): ju.Iterator[Node[K, V]] =
+    new NodeIterator
+
+  private[util] def keyIterator(): ju.Iterator[K] =
+    new KeyIterator
+
+  private[util] def valueIterator(): ju.Iterator[V] =
+    new ValueIterator
+
+  // The cast works around the lack of definition-site variance
+  private[util] final def entrySetIterator(): ju.Iterator[Map.Entry[K, V]] =
+    nodeIterator().asInstanceOf[ju.Iterator[Map.Entry[K, V]]]
+
+  private final class NodeIterator extends AbstractHashMapIterator[Node[K, V]] {
+    protected[this] def extract(node: Node[K, V]): Node[K, V] = node
+  }
+
+  private final class KeyIterator extends AbstractHashMapIterator[K] {
+    protected[this] def extract(node: Node[K, V]): K = node.key
+  }
 
-    override def iterator(): Iterator[V] = {
-      new AbstractMapViewIterator[V] {
-        protected def getNextForm(key: Box[K]): V = inner(key)
+  private final class ValueIterator extends AbstractHashMapIterator[V] {
+    protected[this] def extract(node: Node[K, V]): V = node.value
+  }
+
+  private abstract class AbstractHashMapIterator[A] extends ju.Iterator[A] {
+    private[this] val len = table.length
+    private[this] var nextIdx: Int = _
+    private[this] var nextNode: Node[K, V] = _
+    private[this] var lastNode: Node[K, V] = _
+
+    protected[this] def extract(node: Node[K, V]): A
+
+    def hasNext(): Boolean = {
+      // scalastyle:off return
+      if (nextNode ne null) {
+        true
+      } else {
+        while (nextIdx < len) {
+          val node = table(nextIdx)
+          nextIdx += 1
+          if (node ne null) {
+            nextNode = node
+            return true
+          }
+        }
+        false
       }
+      // scalastyle:on return
+    }
+
+    def next(): A = {
+      if (!hasNext())
+        throw new NoSuchElementException("next on empty iterator")
+      val node = nextNode
+      lastNode = node
+      nextNode = node.next
```

Comment:
I find it a bit odd that going to the next node is split across two methods. My hunch is that it is simpler to deal with initial conditions this way (and look for the first node). So maybe just add a comment explaining that?